### PR TITLE
Add Prometheus Blackbox Exporter

### PR DESCRIPTION
**Pull request by kfox1111**
This PS adds support for using Prometheus Blackbox Exporter

### CODE REVIEW

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for the PR. Please checkout our guidelines and best practices. Make sure you use namespaced partials in `_helpers.tpl`. Please compare with what `helm create` with Helm 2.8+ gives you and update accordingly.

**Comment by kfox1111** [comment]
hmm.... would have thought the checker would have checked for that... Do you know where the code is for the circleci checkers?

**Comment by unguiculus** [comment]
Here we go: https://github.com/kubernetes/charts/tree/master/test/circle

File: stable/blackbox-exporter/templates/deployment.yaml
Comment (lswith) [inline stable/blackbox-exporter/templates/deployment.yaml]: You've used livenessProbe twice here.

**Comment by vglafirov** [comment]
Any progress on it?

**Comment by davidewatson** [comment]
I am also interested in this. I started writting my own chart last week since I need this functionality relatively soon.

**Comment by kfox1111** [comment]
I probably wont have time to work on it for the next few days. @davidewatson, do you have time to port this to a freshly 'helm create''ed directory in the next few days?

**Comment by davidewatson** [comment]
@kfox1111: Sure thing. Is there a reason to start with a "freshly helm created directory" instead of submitting a PR against your branch or opening another PR with your work?

**Comment by kfox1111** [comment]
PR against this branch is good. The review above just wanted to make sure we started from a 'helm create'd version. The _helpers.yaml has the new default goodness. Thanks.

**Comment by davidewatson** [comment]
https://github.com/kfox1111/charts/pull/1 ports this to the `_helpers.tpl` generated by helm 2.8... I'll check back tomorrow to see if any more changes are necessary. Once this is merged, I would like to port it to `coreos/prometheus-operator`.

**Comment by kfox1111** [comment]
@unguiculus I think the requested changes are in place. Please re-review. Thanks. :)

**Comment by timfpark** [comment]
Its not clear to me exactly what configuration needs to occur within this chart vs. an existing Prometheus helm deployment to use this helm chart for monitoring an endpoint.  It would probably save a lot of struggling by a lot of people to figure this out if you added a simple step by step example on how to configure this directly here.

**Comment by kfox1111** [comment]
@timfpark I agree, except that the instructions for integrating it with the prometheus chart are likely to be different then with the prometheus-operator. I think better documentation should not hold up getting this out there? Better docs could come as a follow on patch set.

Also, a follow on patch set adding this chart as a optional dep to the prometheus chart is also another way to handle it.

File: stable/blackbox-exporter/templates/configmap.yaml
Comment (unguiculus) [inline stable/blackbox-exporter/templates/configmap.yaml]: Please update `chart` labels as well.

`chart: {{ template "blackbox-exporter.chart" . }}`

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by andrein** [comment]
I have signed the CLA.

**Comment by simontesar** [comment]
Is there anything I can be of help with to get this chart merged? It shows that there still changes requested but those seem to have been fixed.

File: stable/blackbox-exporter/templates/deployment.yaml
Comment (unguiculus) [inline stable/blackbox-exporter/templates/deployment.yaml]: Instead of making this configurable we should just go with `apps/v1beta2`.

File: stable/blackbox-exporter/templates/configmap.yaml
Comment (unguiculus) [inline stable/blackbox-exporter/templates/configmap.yaml]: I'd just do this instead:

`{{- if .Values.config }}`

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/blackbox-exporter/templates/deployment.yaml
Comment (kfox1111) [reply stable/blackbox-exporter/templates/deployment.yaml]: -1. There are 3 versions in active use apps/v1beta2, apps/v1beta1 and apps/v1. I have a production cluster I'm running this chart on that does not support v1beta2+, only v1beta1. I think its 1.9+ that recommends no beta at all or is that 1.10? We should set it to a good default but let it be configurable for a while, until all potential users get up to a more modern version.

File: stable/blackbox-exporter/templates/configmap.yaml
Comment (kfox1111) [reply stable/blackbox-exporter/templates/configmap.yaml]: ok.

File: stable/blackbox-exporter/templates/deployment.yaml
Comment (prydonius) [reply stable/blackbox-exporter/templates/deployment.yaml]: We officially support N and N-1 (https://github.com/kubernetes/charts#supported-kubernetes-versions), so we only need to use the latest version that works on 1.8. According to the docs, Deployments are apps/v1beta2 there (https://v1-8.docs.kubernetes.io/docs/api-reference/v1.8/).

File: stable/blackbox-exporter/README.md
Comment (gianrubio) [inline stable/blackbox-exporter/README.md]: this is not the right image repository

File: stable/blackbox-exporter/README.md
Comment (gianrubio) [inline stable/blackbox-exporter/README.md]: should be 11.0

File: stable/blackbox-exporter/values.yaml
Comment (gianrubio) [inline stable/blackbox-exporter/values.yaml]: Why do you have this proxy image configured here? Rather than expose trough nginx you should create an ingress object and allow the users to customise it

Comment (gianrubio) [reply stable/blackbox-exporter/values.yaml]: If you want to protect this resource either you use `networkPolicies` or a proxy like oauth configured[ trough annotations](https://github.com/kubernetes/ingress-nginx/tree/master/docs/examples/external-auth)

**Comment by gianrubio** [comment]
this chart should be renamed to postgres-blackbox-exporter

File: stable/blackbox-exporter/values.yaml
Comment (kfox1111) [reply stable/blackbox-exporter/values.yaml]: The resource is protected by password inside the cluster.

**Comment by kfox1111** [comment]
@gianrubio I'm going to assume you meant prometheus. I'll rename it to that.

File: stable/blackbox-exporter/values.yaml
Comment (kfox1111) [reply stable/blackbox-exporter/values.yaml]: You can also use networkPolicies or oauth proxies with this chart. The use of the fronting nginx is optional. For sites that want to spread around blackbox exporters so they can increase their visibility, all they need is this chart set to net=host and k8s/helm on a single node. No ingress needed. Monitoring should be as simple as possible.

File: stable/blackbox-exporter/templates/deployment.yaml
Comment (kfox1111) [reply stable/blackbox-exporter/templates/deployment.yaml]: Ok. I'll just keep a fork around for my old cluster. it should be going away in a month or so anyway.

File: stable/blackbox-exporter/values.yaml
Comment (gianrubio) [reply stable/blackbox-exporter/values.yaml]: I still discourage you to run a sidecar to protect blackbox, again: you can achieve the same results with ingress and networkPolicies. To me it's an anti pattern because k8s already solved this problem for you.

Comment (kfox1111) [reply stable/blackbox-exporter/values.yaml]: Generally I agree. but Monitoring is an entirely different animal then most. You want your monitoring system to be as simple as possible, otherwise you have to monitor your monitoring systems monitoring system and its turtles all the way down. With using the "normal" k8s stuff, you have to ensure you are using SDN. Which is a big complicated animal. The simpliest ones, such as flannel doesn't even support networkPolicies. Then you need to layer not only nginx but an ingress controller on top of that. Just to ping a http request off of another box. (normally I'm all for the SDN stuff. its great)

We're talking about:
k8s + helm + canal/weave + service + ingress + ingress-controller + externalip/nodeport + kube-proxy + complicated iptables rulesets + blackbox-exporter
Which is WAY more complicated then:
k8s + helm + (nginx + blackbox-exporter, pod with net=host)

The goal of this particular chart, not charts in general is simplicity. I think it is exception worthy.

**Comment by kfox1111** [comment]
@unguiculus please review

File: stable/prometheus-blackbox-exporter/templates/deployment.yaml
Comment (unguiculus) [inline stable/prometheus-blackbox-exporter/templates/deployment.yaml:1]: Isn't that rather complicated? We only need a single variable.

File: stable/prometheus-blackbox-exporter/templates/deployment.yaml
Comment (unguiculus) [inline stable/prometheus-blackbox-exporter/templates/deployment.yaml:94]: `{{- if .Values.config }}`

File: stable/prometheus-blackbox-exporter/templates/deployment.yaml
Comment (unguiculus) [inline stable/prometheus-blackbox-exporter/templates/deployment.yaml:99]: `{{- if .Values.config }}`

File: stable/prometheus-blackbox-exporter/templates/deployment.yaml
Comment (unguiculus) [inline stable/prometheus-blackbox-exporter/templates/deployment.yaml:104]: Add space before `}}`

File: stable/blackbox-exporter/values.yaml
Comment (unguiculus) [reply stable/blackbox-exporter/values.yaml]: I agree with @gianrubio. The proxy should not be part of the chart. I'd rather suggest you make this configurable. Here's an example:

https://github.com/kubernetes/charts/blob/86c217302513fed924cfd57808fbc9b51c031623/stable/nginx-ingress/values.yaml#L187-L219

**Comment by unguiculus** [comment]
I also commented above which might be hidden: https://github.com/kubernetes/charts/pull/3800#discussion_r185048276

File: stable/blackbox-exporter/values.yaml
Comment (prydonius) [reply stable/blackbox-exporter/values.yaml]: I understand what @kfox1111 is saying here, but I do also think this is a bit of an anti-pattern. Based on your argument about keeping it simple, what's the point of running monitoring systems on Kubernetes at all, when you could just run something directly on the host VMs?

That said, given that this is disabled by default and not a very complex configuration, I think it may be _okay_. I would just add comments above the value here and in the README to explain exactly what and how this proxyImage is used (e.g. only used if you set a password).

I think @unguiculus' approach is a good one too, you could always include a separate values.yaml file that includes the nginx basic auth example.

**Comment by prydonius** [comment]
Also incase it was missed: https://github.com/kubernetes/charts/pull/3800#discussion_r185198813

**Comment by gianrubio** [comment]
@kfox1111 do you have time to fix the changes  requested?

**Comment by kfox1111** [comment]
No. I think we're at an impasse. Honestly, the requested changes around nginx make it unusable to me without basically forking just about the entire chart anyway. So someone else should probably carry it if that is the path forward.

File: stable/blackbox-exporter/templates/deployment.yaml
Comment (davidewatson) [reply stable/blackbox-exporter/templates/deployment.yaml]: Completely agree that forcing the user to set/see the `apiVersion` may not be ideal. Afaik, I only care that the right one is chosen for a given k8s cluster...

Since helm [2.2.0](https://github.com/kubernetes/helm/releases/tag/v2.2.0) it is also possible to check for capabilities. `prometheus-operator` [does this](https://github.com/coreos/prometheus-operator/search?utf8=%E2%9C%93&q=Capabilities.APIVersions&type=) for RBAC. Maybe we should adopt this pattern too.

**Comment by estahn** [comment]
Is this pull request dead? Should we create a new one to get this through?

**Comment by gianrubio** [comment]
@estahn see this PR https://github.com/kubernetes/charts/pull/6134

**Comment by bhack** [comment]
It was merged at https://github.com/kubernetes/charts/pull/6134 but it still has some small bugs.

**Comment by desaintmartin** [comment]
Are the bugs documented somewhere? I don't see them (at least list of remaining issues) in the PRs.

**Pull request closed without merging**
